Fix usage of fs.listdir in CheckpointConnector

leoleoasd opened the pull request:
## What does this PR do?

<!--
Please include a summary of the change and which issue is fixed.
Please also include relevant motivation and context.
List any dependencies that are required for this change.

If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

The following links the related issue to the PR (https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
-->

Fix use path in fs.listdir in checkpoint

Fixes #15379

### Does your PR introduce any breaking changes? If yes, please list them.

<!-- FILL IN or None -->
no

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

<!-- In the CHANGELOG, separate each item in the unreleased section by a blank line to reduce collisions -->

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

leoleoasd commented:
Only fixing `listdir` for now. Other usages of `fs.xxxx` still need to be discussed

awaelchli commented on src/pytorch_lightning/trainer/connectors/checkpoint_connector.py:
```diff
@@ -574,12 +575,12 @@ def __max_ckpt_version_in_folder(dir_path: _PATH, name_key: str = "ckpt_") -> Op
 
         # check directory existence
-        fs = get_filesystem(dir_path)
+        fs, uri = url_to_fs(dir_path)
```
@leoleoasd why do we have to `fs.listdir(uri)` but` fs.exists(dir_path)` is ok?
There are many other places in our code base where we use fsspec (modelcheckpiont, saving utilities, tensorboard, profiler, etc.). Does that mean we need to update the usage there too?

leoleoasd replied:
https://filesystem-spec.readthedocs.io/en/latest/api.html#fsspec.spec.AbstractFileSystem
According to the document, yes. 
But for mysterious reasons, `exists` works but `listdir` don't work. I'm only changing `listdir` because it blocks my work, and I'm not sure if we have to follow the standard behavior or "if it works, don't touch it". Do we have unit tests for fsspec file systems?

In the document, all functions that aren't bound to a specific FS have parameters named `url` while all methods of `AbstractFileSystem` have parameters named `path`. I think we should update all usages (but not for tensorboard, since tensorboard has its own way of calling fsspec)

leoleoasd commented:
Hi @Borda @awaelchli any update on the review of this PR?

carmocca reviewed (commented):
There are several failing tests.

Also, a CHANGELOG entry should be added.

leoleoasd commented:
Do we actually have any tests for remote file systems? Did anyone try running lightning apps locally with `file://` or `local://`? I don't see any tests in #3805 and #2164.

![image](https://user-images.githubusercontent.com/37735580/199662990-6cf2a447-c09e-4647-9c00-2d5a117f3606.png)
There aren't any string that contains protocol other than HTTP or HTTPS in all PyTorch tests (apart from what I just added).

I added a test that uses fsspec's in-memory file system, and other things related to the test (_hpc_resume_path) are broken too.
I think we DO need to change EVERY usage of fsspec.

awaelchli commented on tests/tests_pytorch/trainer/connectors/test_checkpoint_connector.py:
```diff
@@ -102,6 +102,28 @@ def test_hpc_max_ckpt_version(tmpdir):
+
+    model = BoringModel()
+    # hardcoding dir since `tmpdir` can be windows path
+    trainer = Trainer(default_root_dir="memory://test_ckpt_for_fsspec", max_steps=1)
+    trainer.fit(model)
```
```suggestion
    trainer = Trainer(default_root_dir="memory://test_ckpt_for_fsspec", limit_train_batches=1, limit_val_batches=1)
    trainer.fit(model)
```

awaelchli commented:
@leoleoasd Responding to your concerns. Yes, we want to support the remote filesystems. It's been added in some places of the code base, logger outputs and checkpointing are the main areas where this is supported. There was never a huge demand, but the users who do use it are mostly using the s3 protocol (from general feedback on slack). For your use cases, you did the right thing and request it on GitHub. Thanks for providing help in improving the coverage here!

carmocca commented on tests/tests_pytorch/trainer/connectors/test_checkpoint_connector.py:
```diff
@@ -102,6 +102,28 @@ def test_hpc_max_ckpt_version(tmpdir):
+    trainer = Trainer(default_root_dir="memory://test_ckpt_for_fsspec", max_steps=1)
+    trainer.fit(model)
+    trainer.save_checkpoint("memory://test_ckpt_for_fsspec/hpc_ckpt.ckpt")
+    trainer.save_checkpoint("memory://test_ckpt_for_fsspec/hpc_ckpt_0.ckpt")
```
Windows jobs are not happy about these slashes

leoleoasd replied:
I do that to fix tests on windows. fsspec wants / as sep on all platforms.

leoleoasd replied:
I don't have windows testing environment,so I have to push and run tests on GitHub actions. This change is to make fsspec works on windows (we don't want \ in s3 or hdfs paths), but it seems there is another error. I'll fix that later

leoleoasd commented:
Tests passed, this PR is ready for the next round of review.
The checkpointing issue is fixed, and I'll open new PR if I found some other things that doesn't work with HDFS and add necessary tests.

awaelchli commented on src/pytorch_lightning/trainer/connectors/checkpoint_connector.py:
```diff
@@ -59,13 +61,16 @@ def __init__(self, trainer: "pl.Trainer", resume_from_checkpoint: Optional[_PATH
         max_version = self.__max_ckpt_version_in_folder(dir_path_hpc, "hpc_ckpt_")
         if max_version is not None:
-            return os.path.join(dir_path_hpc, f"hpc_ckpt_{max_version}.ckpt")
+            if isinstance(fs, LocalFileSystem):
+                return os.path.join(dir_path_hpc, f"hpc_ckpt_{max_version}.ckpt")
+            else:
+                return dir_path_hpc + fs.sep + f"hpc_ckpt_{max_version}.ckpt"
```
I wonder if we will need a utility function for this in the future (I feel like fsspec should have a "join" function). Let's keep it in mind.

leoleoasd replied:
According to fsspec/filesystem_spec#747, they don't have that yet but it's coming soon

carmocca reviewed (approved):
LGTM if CI agrees
